src/aggregator/client/queue.go

```diff
@@ -204,22 +209,61 @@ func (q *queue) Close() error {
 	return nil
 }
 
+func (q *queue) writeAndReset() {
+	if len(q.buf) == 0 {
+		return
+	}
+	if err := q.writeFn(q.buf); err != nil {
+		q.log.Error("error writing data on timed flush",
+			zap.String("instance", q.instance.Endpoint()),
+			zap.Error(err),
+		)
+		q.metrics.connWriteErrors.Inc(1)
+	} else {
+		q.metrics.connWriteSuccesses.Inc(1)
+	}
+	q.buf = q.buf[:0]
+}
+
 func (q *queue) drain() {
 	defer q.wg.Done()
+	defer q.conn.Close()
+	timer := time.NewTimer(q.batchFlushDeadline)
```

Comment:
@prateek I think Prateek can chime in on this but apparently its better to use `time.Sleep` because they have some runtime integration which makes them more likely to fire on time as opposed to timers (which can get super delayed)

Reply:
I'm wondering whether it's worth changing this to use sleeps - I'll have to add another goroutine (and we've got a bunch of them already). On any reasonably active connection we'll still be flushing in the normal write path based on time since last drain.